### DISCUSSION TITLE
qb.execute doesn't do mapping of fields for joined entities

### DISCUSSION THREAD

**Discussion by dirslashls** [mikro-orm/mikro-orm#4726] [General]
**Describe the bug**

In the following query, the `created_date` column ends up as `createdDate` in the returned JSON as expected. However, the `isActive` property shows up as `is_active`.

```ts
em.qb(TeamMember,'tm')
  .leftJoin('tm.member','u')
  .select(['tm.id','tm.createdDate','u.name','u.isActive'])
  .execute('all');
```

**To Reproduce**
Setup two entities with camel case properties that can be joined. Use query builder to join and the `execute` method. Notice that only the main entity fields are mapped.

**Expected behavior**
I was expecting all selected fields to be mapped to their property names.

**Versions**

| Dependency | Version |
| - | - |
| node | v16.15.0 |
| typescript | Version 4.7.4 |
| mikro-orm | 5.8.1 |
| your-driver | "@mikro-orm/mysql": "^5.8.1" |

**Comment by B4nan** [comment]
What do your entities look like? And your ORM config?

**Reply by dirslashls** [reply]
Thanks @B4nan for looking into this. Here is a simple test case with entities.

testa.ts

```ts
import {  Entity, PrimaryKey, Property } from '@mikro-orm/core';

@Entity()
class TestA {
	@PrimaryKey()
	id: number;

	@Property()
	isActive: boolean
}

export { TestA };
```

testb.ts

```ts
import { Entity, ManyToOne, PrimaryKey, Property, Ref } from '@mikro-orm/core';
import { TestA } from './testa.js';

@Entity()
class TestB {
	@PrimaryKey()
	id: string;

	@Property()
	createdDate: Date;

	@Property()
	isActive2: boolean;

	@ManyToOne('TestA',{ cascade: [] })
	a?: Ref<TestA>;
}

export { TestB };
```

The following code

```ts
const qb = em.qb(TestB,'tm')
		.leftJoin('tm.a','u')
		.select(['tm.id','tm.createdDate','tm.isActive2','u.name','u.isActive']);
console.debug(qb.getQuery());
const data = await qb.execute('all');
console.debug(data);
```

displays the following in the log

```txt
select `tm`.`id`, `tm`.`created_date`, `tm`.`is_active2`, `u`.`name`, `u`.`is_active` from `test_b` as `tm` left join `test_a` as `u` on `tm`.`a_id` = `u`.`id`
[query] select `tm`.`id`, `tm`.`created_date`, `tm`.`is_active2`, `u`.`name`, `u`.`is_active` from `test_b` as `tm` left join `test_a` as `u` on `tm`.`a_id` = `u`.`id` [took 8 ms, 1 result]
[
  {
    id: '1',
    createdDate: 2023-09-21T00:00:00.000Z,
    isActive2: true,
    name: 'Test 123',
    is_active: 1
  }
]
```

Notice how isActive2 property of the main entity of the qb maps fine. ORM configuration is as follows where dbconfig is just the database login details. BTW, I also just noticed `u.isActive` not only shows up as `is_active`, the value is `1` instead of `true`.

```ts
	const config = { ...dbconfig,
		timezone: '+00:00',
		forceUtcTimezone: true,
		entities: [path.join(__dirname,'../dist/entity/**/*.js')],
		entitiesTs: [path.join(__dirname,'../src/entity/**/*.ts')],
		metadataProvider: TsMorphMetadataProvider,
		discovery: { disableDynamicFileAccess: isProd },
		debug,
		dynamicImportProvider: (id: string) => import(id),
		autoJoinOneToOneOwner: false,
		schemaGenerator: {
			disableForeignKeys: true,
			createForeignKeyConstraints: false,
		},
}
```

**Reply by B4nan** [reply]
Oh now I get it, you are mixing multiple entity types in the select clause. This will indeed map only the root entity properties, the result mapping layer works based on entity types, it's not a "per property mapping" or how to call it. To be honest I am not sure how to fix this now... Will try to do something about it in v6, but first have to finish with some WIP stuff.

Generally speaking, there is the problem of aliasing, as you just select bunch of things from different types and they could shadow themselves - in other words, that `isActive` property might have lived on the root entity as well, and in that case, we don't know where to look. For this problem, the QB offers the `joinAndSelect` methods that automatically alias things, so I think that will be required for any fix to exist:

```ts
em.qb(TeamMember,'tm')
  .select(['tm.id', 'tm.createdDate'])
  .leftJoinAndSelect('tm.member','u', {}, ['name', 'isActive'])
  .execute('all');
```

(this itself won't help right now, it will just alias the joined columns, but that will be required for the mapping to work correctly once we fix that)

And thinking about it more, better solution would be having a way to alias the fields explicitly, otherwise we will need to do a some nasty regexp stuff on the way. Overall this will be a major refactor and as such, I will try to do it in v6.

**Reply by dirslashls** [reply]
I just tried the joinAndSelect and while it solves the property mapping, it still has two issues:

1. The result is now nested while I wanted a flat result.
```js
[
  {
    id: '1',
    createdDate: 2023-09-22T00:00:00.000Z,
    isActive2: true,
    a: { id: 1, name: 'Test 123', isActive: 1 }
  }
]
```

I think the main issue is with the return type of `execute` which assumes it will be the same shape as the root entity. Perhaps a flat select mode (with explicit Aliasing by developer to resolve same properties from different entities) might solve this. I don't know if it will be possible to still have a strong typing at that point to the return value.

2. The nested isActive is still returned as `1` instead of as `true`.

Whatever timeline you choose to address these is fine, I can workaround these. Thank you.
